File: src/com/chbachman/cminus/representation/Scope.java
package com.chbachman.cminus.representation;

import com.chbachman.cminus.representation.function.CodeBlock;
import com.chbachman.cminus.representation.function.Function;
import com.chbachman.cminus.representation.value.Value;
import com.chbachman.cminus.representation.value.Variable;
import com.sun.tools.javac.jvm.Code;

import java.util.*;

/**
 * Created by Chandler on 4/12/17.
 */
public class Scope implements Iterable<Scope.ScopeHolder>{

    ScopeHolder head = new ScopeHolder();

    public CodeBlock popScope() {
        ScopeHolder current = this.head;
        Optional<ScopeHolder> parent = head.parent;
        if(parent.isPresent()) {
            this.head = parent.get();
        }

        return current.block;
    }

    public void pushScope(CodeBlock block) {
        head = new ScopeHolder(head, block);
        block.setupScope(this);
    }

    public CodeBlock code() {
        return head.block;
    }

    public void addVariable(Variable v) {
        if (v == null) {
            return;
        }
        System.out.println(head.vars);
        head.vars.put(v.name, v);
        System.out.println(head.vars);
    }

    public Optional<Variable> getVariable(String name) {
        for (ScopeHolder scope : this) {
            if (scope.vars.containsKey(name)) {
                return Optional.of(scope.vars.get(name));
            }
        }

        return Optional.empty();
    }

    public void addFunction(Function f) {
        if (f == null) {
            return;
        }

        if (getFunction(f.getBaseName(), f.parameters).isPresent()) {
            throw new RuntimeException("The function " + f.getBaseName() + " already exists.");
        }

        if (head.functions.containsKey(f.getBaseName())) {
            head.functions.get(f.getBaseName()).add(f);
            return;
        }

        List<Function> newList = new ArrayList<>();
        newList.add(f);

        head.functions.put(f.getBaseName(), newList);
    }

    public void addStruct(Struct s) {
        if (s == null) {
            return;
        }

        head.structs.put(s.name, s);
    }

    public List<Struct> getStructs() {
        List<Struct> structList = new ArrayList<>();

        for(ScopeHolder s : this) {
            structList.addAll(s.structs.values());
        }

        return structList;
    }

    public Optional<Struct> getStruct(String name) {
        for (ScopeHolder scope : this) {
            if (scope.structs.containsKey(name)) {
                return Optional.of(scope.structs.get(name));
            }
        }

        return Optional.empty();
    }

    public boolean functionNameExists(String name) {
        for (ScopeHolder scope: this) {
            if (scope.functions.containsKey(name)) {
                return true;
            }
        }

        return false;
    }

    public Optional<Function> getFunction(String name, List<? extends Typed> parameters) {
        for (ScopeHolder scope : this) {
            if (scope.functions.containsKey(name)) {
                List<Function> functions = scope.functions.get(name);

                loop:
                for (Function f : functions) {
                    if (f.parameters.size() != parameters.size()) {
                        continue loop;
                    }

                    for (int i = 0; i < f.parameters.size(); i++) {
                        if (parameters.get(i).type() != f.parameters.get(i).type()) {
                            continue loop;
                        }
                    }

                    return Optional.of(f);
                }
            }
        }

        return Optional.empty();
    }

    @Override
    public Iterator<ScopeHolder> iterator() {
        return new ScopeIterator(this);
    }

    private class Main implements CodeBlock {
        public Main() {}

        @Override
        public String first() {
            return null;
        }

        @Override
        public String middle() {
            return null;
        }

        @Override
        public String last() {
            return null;
        }
    }

    private class ScopeIterator implements Iterator<ScopeHolder> {

        Optional<ScopeHolder> scope;

        ScopeIterator(Scope s) {
            this.scope = Optional.of(s.head);
        }

        @Override
        public boolean hasNext() {
            return scope.isPresent();
        }

        @Override
        public ScopeHolder next() {
            ScopeHolder s = scope.get();
            scope = s.parent;
            return s;
        }
    }


    class ScopeHolder {
        Map<String, Variable> vars = new TreeMap<>();
        Map<String, List<Function>> functions = new TreeMap<>();
        Map<String, Struct> structs = new TreeMap<>();

        Optional<ScopeHolder> parent;
        final CodeBlock block;

        public ScopeHolder() {
            this(null, new Main());
        }

        public ScopeHolder(ScopeHolder parent, CodeBlock block) {
            this.parent = Optional.ofNullable(parent);
            this.block = block;
        }
    }
}

File: src/main/java/com/chbachman/cminus/representation/Scope.kt
package com.chbachman.cminus.representation

import com.chbachman.cminus.representation.function.CodeBlock
import com.chbachman.cminus.representation.function.Header
import com.chbachman.cminus.representation.struct.StructHeader
import com.chbachman.cminus.representation.value.Identifier
import com.chbachman.cminus.representation.value.Variable
import com.chbachman.cminus.util.peek
import com.chbachman.cminus.util.pop
import com.chbachman.cminus.util.push

/**
 * Created by Chandler on 4/12/17.
 * Handles the declaration of variables. Uses a stack to represent the current scope.
 * Also handles structs and function declarations, but on a global scope.
 */
class Scope(functions: List<Header>, structs: List<StructHeader>) {

    private val stack = mutableListOf<ScopeHolder>()
    val functions = functions.groupBy { it.baseName }

    // Set the "this" containing variable.
    fun setThis(variable: Variable, container: VariableHolder) {
        stack.peek()?.thisVar = Pair(variable, container)
    }

    // Remove the current scope.
    fun popScope(): CodeBlock {
        return stack.pop()?.block ?: throw RuntimeException("Could not pop empty stack.")
    }

    // Add to scope.
    fun pushScope(block: CodeBlock) {
        stack.push(ScopeHolder(block))
        block.setupScope(this)
    }

    fun addVariable(v: Variable) {
        stack.peek()?.vars?.put(v.name.last.text, v)
    }

    fun getVariable(id: Identifier): Variable? {
        // The initial variable name
        val first = handleAllButLast(id) ?: return null

        return first + (id.last.getVar(first) ?: return null)
    }

    private fun handleAllButLast(id: Identifier): Variable? {
        val segments = id.segments.dropLast(1)
        val it = segments.iterator()

        // Get the first var. If we don't have it, then just return.
        val start = getFirstVariable(it) ?: return null

        // If it exhausts the full range, then we just return the self variable.
        if (!it.hasNext()) {
            return start
        }

        return getVariable(it, start)
    }

    private fun getFirstVariable(id: Iterator<Identifier.Segment>): Variable? {
        if (!id.hasNext()) {
            return null
        }

        val name = run {
            val name = id.next().text

            if (name == "self") {
                if (!id.hasNext()) {
                    return stack.peek()?.thisVar?.first
                }

                id.next().text
            } else {
                name
            }
        }

        val self = name == "self"

        for (scope in stack) {
            if (!self && scope.vars.containsKey(name)) {
                return scope.vars[name]
            }

            val thisVar = scope.getThis(name)

            if (thisVar != null) {
                return thisVar
            }
        }

        return null
    }

    private tailrec fun getVariable(id: Iterator<Identifier.Segment>, con: Variable): Variable? {
        // Since we don't have another one, we can just return this one.
        if (!id.hasNext()) {
            return con
        }

        val segment = id.next()

        // Have the segment get the variable from this one.
        val variable = segment.getVar(con) ?: return null

        return getVariable(id, variable)
    }

    private inner class ScopeHolder(val block: CodeBlock) {
        var vars: MutableMap<String, Variable> = hashMapOf()
        var thisVar: Pair<Variable, VariableHolder>? = null

        fun getThis(name: String): Variable? {
            return (thisVar?.first ?: return null) +
                    (thisVar?.second?.getVariable(name) ?: return null)
        }
    }
}

operator fun Map<String, List<Header>>.get(name: String, parameters: List<Typed>): Header? {
    val functions = this[name]

    functions ?: return null

    return functions.find { it.matches(name, parameters) }
}

